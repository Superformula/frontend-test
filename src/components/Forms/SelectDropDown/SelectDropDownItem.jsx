import React from 'react';
import PropTypes from 'prop-types';
import { borderColor, borders } from 'styled-system';
import styled from 'styled-components';

import CircleCheck from '../../../assets/circle-check.svg';
import { GridItem } from '../..';

const StyledCheckbox = styled.div`
  display: inline-block;
  width: 16px;
  height: 16px;
  border-radius: 50%;
  ${borders}
  ${borderColor}
`;

const SelectContentsItem = styled(GridItem)`
  outline: none;
  cursor: pointer;
  color: ${props => props.theme.colors.black60}
  &:hover {
    color: ${props => props.theme.colors.primary}
  }
`;

const ItemTitle = styled.span`
  padding-left: 4px;
`;

const SelectDropDownItem = ({ item, selectedItem, onSelect }) => {
  const icon = selectedItem === item.value
    ? <CircleCheck />
    : (
      <StyledCheckbox
        border="1px solid"
        borderColor="black21"
      />
    );
  return (
    <li>
      <SelectContentsItem
        // TODO: remove auto-binding onSelect functions
        onClick={() => onSelect(item)}
        role="button"
        onKeyPress={() => onSelect(item)}
        tabIndex={0}
        display="flex"
        alignItems="center"
        py={2}
      >
        {icon}
        <ItemTitle>{item.title}</ItemTitle>
      </SelectContentsItem>
    </li>
  );
};

SelectDropDownItem.defaultProps = {
  selectedItem: undefined,
};

SelectDropDownItem.propTypes = {
  item: PropTypes.shape({
    title: PropTypes.string,
    value: PropTypes.string,
  }).isRequired,
  onSelect: PropTypes.func.isRequired,
  selectedItem: PropTypes.string,
};

export default SelectDropDownItem;
