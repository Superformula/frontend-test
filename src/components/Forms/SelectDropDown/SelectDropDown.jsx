import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { space } from 'styled-system';
import onClickOutside from 'react-onclickoutside';

import { GridItem } from '../..';
import SelectDropDownItem from './SelectDropDownItem';
import ChevronUp from '../../../assets/chevron-up.svg';
import ChevronDown from '../../../assets/chevron-down.svg';

const ChevronWrapper = styled.span`
  > svg {
    width: 12px;
    height: 12px;
    ${space}
    stroke: ${props => props.theme.colors.black41}
  }
`;

const SelectContents = styled.ul`
  border: 1px solid ${props => props.theme.colors.black21};
  box-shadow: 0 6px 5px rgba(0, 0, 0, 0.1);
  background: ${props => props.theme.colors.white};
  margin-top: -1px;
  padding: 8px 12px;
  list-style: none;
  position: absolute;
  z-index: 1;
  width: 100%;
`;

const DropDownHeader = styled.div`
  display: inline-flex;
  align-items: center;
  justify-content: space-between;
  color: ${props => props.theme.colors.primary}
  border-bottom: 1px solid ${props => props.theme.colors.black21}
  padding-bottom: 4px;
  outline: 0;
  cursor: pointer;
  width: 100%;
`;

class SelectDropDown extends PureComponent {
  state = {
    open: false,
  };

  handleClickOutside = () => {
    this.setState({ open: false });
  };

  onSelect = (selectedItem) => {
    const { onSelect } = this.props;
    onSelect(selectedItem);
    this.toggleDropDown();
  };

  toggleDropDown = () => {
    const { open } = this.state;
    this.setState({ open: !open });
  };

  render() {
    const {
      heading,
      items,
      selectedItem,
    } = this.props;
    const { open } = this.state;
    const chevron = open ? <ChevronUp /> : <ChevronDown />;
    return (
      <GridItem position="relative">
        <DropDownHeader
          onClick={this.toggleDropDown}
          role="button"
          onKeyPress={this.toggleDropDown}
          tabIndex={0}
        >
          {heading}
          <ChevronWrapper>{chevron}</ChevronWrapper>
        </DropDownHeader>
        {open
        && (
          <SelectContents>
            {items.map(item => (
              <SelectDropDownItem
                item={item}
                onSelect={this.onSelect}
                selectedItem={selectedItem}
                key={item.title}
              />
            ))}
          </SelectContents>
        )}
      </GridItem>
    );
  }
}

SelectDropDown.propTypes = {
  heading: PropTypes.string.isRequired,
  items: PropTypes.arrayOf(PropTypes.shape({
    title: PropTypes.string,
    value: PropTypes.string,
  })).isRequired,
  onSelect: PropTypes.func.isRequired,
  selectedItem: PropTypes.string.isRequired,
};

export default onClickOutside(SelectDropDown);
