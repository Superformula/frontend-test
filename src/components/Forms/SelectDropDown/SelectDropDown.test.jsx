import React from 'react';
import { mountWithTheme } from '../../../helpers/tests/withTheme';
import SelectDropDown from './SelectDropDown';

const defaultProps = {
  heading: 'Test',
  items: [{ title: 'Test', value: 'test' }],
  onSelect: jest.fn(),
  selectedItem: 'test',
};

describe('SelectDropDown Component', () => {
  const component = mountWithTheme(<SelectDropDown {...defaultProps}>Primary</SelectDropDown>).find('SelectDropDown');
  it('renders correctly', () => {
    expect(component).toMatchSnapshot();
  });
});
