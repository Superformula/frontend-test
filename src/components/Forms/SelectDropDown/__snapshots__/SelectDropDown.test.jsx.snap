// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SelectDropDown Component renders correctly 1`] = `
<SelectDropDown
  disableOnClickOutside={[Function]}
  enableOnClickOutside={[Function]}
  eventTypes={
    Array [
      "mousedown",
      "touchstart",
    ]
  }
  heading="Test"
  items={
    Array [
      Object {
        "title": "Test",
        "value": "test",
      },
    ]
  }
  onSelect={[MockFunction]}
  outsideClickIgnoreClass="ignore-react-onclickoutside"
  preventDefault={false}
  selectedItem="test"
  stopPropagation={false}
>
  <styled.div
    position="relative"
  >
    <div
      className="sc-gqjmRU bAkkMC"
    >
      <styled.div
        onClick={[Function]}
        onKeyPress={[Function]}
        role="button"
        tabIndex={0}
      >
        <div
          className="sc-ckVGcZ cnuynl"
          onClick={[Function]}
          onKeyPress={[Function]}
          role="button"
          tabIndex={0}
        >
          Test
          <styled.span>
            <span
              className="sc-kpOJdX iFRxyn"
            >
              <SvgChevronDown>
                <svg
                  data-file-name="SvgChevronDown"
                />
              </SvgChevronDown>
            </span>
          </styled.span>
        </div>
      </styled.div>
    </div>
  </styled.div>
</SelectDropDown>
`;
